Source/WebKit/Shared/WebBackForwardListFrameItem.cpp

```diff
@@ -158,11 +158,13 @@ Ref<FrameState> WebBackForwardListFrameItem::copyFrameStateWithChildren()
     return frameState;
 }
 
-bool WebBackForwardListFrameItem::hasAncestorFrame(FrameIdentifier frameID)
+bool WebBackForwardListFrameItem::sharesAncestor(WebBackForwardListFrameItem& frameItem) const
 {
-    for (RefPtr ancestor = m_parent.get(); ancestor; ancestor = ancestor->m_parent.get()) {
-        if (ancestor->frameID() == frameID)
-            return true;
+    for (RefPtr currentAncestor = m_parent.get(); currentAncestor; currentAncestor = currentAncestor->m_parent.get()) {
+        for (RefPtr frameItemAncestor = frameItem.m_parent.get(); frameItemAncestor; frameItemAncestor = frameItemAncestor->m_parent.get()) {
```

Comment:
This is a classic O(n^2) algorithm.  Could we instead make two sets of identifiers and check the intersection?  That would be O(n)

Reply:
Sure